sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs

```diff
@@ -76,41 +98,129 @@ private static void CheckForRedundantConditions(ExplodedGraph explodedGraph, Syn
             }
         }
 
-        private static void ProcessVisitedBlocks(HashSet<SyntaxNode> conditionTrue, HashSet<SyntaxNode> conditionFalse, SyntaxNodeAnalysisContext context)
+        private static Diagnostic GetDiagnostics(SyntaxNode constantNode, bool constantValue)
+        {
+            var unreachableLocations = GetUnreachableLocations(constantNode, constantValue)
+                .ToList();
+
+            var constantText = constantValue.ToString().ToLowerInvariant();
+
+            return unreachableLocations.Count > 0 ?
+                Diagnostic.Create(s2583, constantNode.GetLocation(), messageArgs: constantText, additionalLocations: unreachableLocations) :
+                Diagnostic.Create(s2589, constantNode.GetLocation(), messageArgs: constantText);
+        }
+
+        private static IEnumerable<Location> GetUnreachableLocations(SyntaxNode constantExpression, bool constantValue)
+        {
+            var unreachabExpressions = GetUnreachableExpressions(constantExpression, constantValue)
+                .ToList();
+
+            if (unreachabExpressions.Count > 0)
+            {
+                yield return Location.Create(
+                    constantExpression.SyntaxTree,
+                    GetSpan(unreachabExpressions.First(), unreachabExpressions.Last()));

```

Comment:
We saw that the indexer access was faster than `.Last()` so you should probably use it.